File: cross-ipc/cross-ipc-go/cross_ipc/cross_ipc.go

package cross_ipc

import (
	"fmt"
	"sync"
	"syscall"
	"unsafe"
)

// ShmDispenserMode represents the mode of operation for a shared memory dispenser
type ShmDispenserMode int

const (
	
	FIFO ShmDispenserMode = 0
	// LIFO mode: Last In, First Out (stack)
	LIFO ShmDispenserMode = 1
	// DEQUE mode: Double-ended queue (can add/remove from both ends)
	DEQUE ShmDispenserMode = 2
)

var (
	dll      *syscall.DLL
	dllMutex sync.Mutex
	dllPath  = "C:/GIthubProjects/cross-ipc/cross-ipc/x64/Debug/cross-ipc.dll" // Default path
)


func SetDLLPath(path string) {
	dllMutex.Lock()
	defer dllMutex.Unlock()
	dllPath = path
}


func loadDLL() (*syscall.DLL, error) {
	dllMutex.Lock()
	defer dllMutex.Unlock()

	if dll != nil {
		return dll, nil
	}

	var err error
	dll, err = syscall.LoadDLL(dllPath)
	if err != nil {
		return nil, fmt.Errorf("failed to load DLL: %w", err)
	}

	return dll, nil
}


func stringToBytes(s string) []byte {
	return []byte(s + "\x00")
}


func ptrToString(ptr uintptr) string {
	if ptr == 0 {
		return ""
	}

	var byteSlice []byte
	for i := 0; ; i++ {
		b := *(*byte)(unsafe.Pointer(ptr + uintptr(i)))
		if b == 0 {
			break
		}
		byteSlice = append(byteSlice, b)
	}

	return string(byteSlice)
}

// Helper function to handle Windows "success" errors
func handleWindowsError(err error) error {
	if err != nil && err.Error() == "The operation completed successfully." {
		return nil
	}
	return err
}

File: cross-ipc/cross-ipc-cpp/src/cross_ipc.cpp
#include "cross_ipc.hpp"
#include <mutex>

namespace cross_ipc {


static std::string dllPath = "C:/GIthubProjects/cross-ipc/cross-ipc/x64/Debug/cross-ipc.dll";
static HMODULE dllHandle = nullptr;
static std::mutex dllMutex;

void SetDLLPath(const std::string& path) {
    std::lock_guard<std::mutex> lock(dllMutex);
    dllPath = path;
    
    
    if (dllHandle) {
        FreeLibrary(dllHandle);
        dllHandle = nullptr;
    }
}

HMODULE LoadDLL() {
    std::lock_guard<std::mutex> lock(dllMutex);
    
    if (dllHandle) {
        return dllHandle;
    }
    
    dllHandle = LoadLibraryA(dllPath.c_str());
    if (!dllHandle) {
        throw CrossIPCError("Failed to load DLL: " + GetLastErrorAsString());
    }
    
    return dllHandle;
}

std::string GetLastErrorAsString() {
    DWORD error = GetLastError();
    if (error == 0) {
        return "";
    }
    
    LPSTR messageBuffer = nullptr;
    size_t size = FormatMessageA(
        FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
        nullptr, error, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPSTR)&messageBuffer, 0, nullptr);
    
    std::string message(messageBuffer, size);
    LocalFree(messageBuffer);
    
    return message;
}

std::string PtrToString(const char* ptr) {
    if (!ptr) {
        return "";
    }
    return std::string(ptr);
}

} 